subprojects/publish/src/main/java/org/gradle/api/publish/internal/mapping/ComponentDependencyResolver.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.publish.internal.mapping;
+
+import org.gradle.api.artifacts.DependencyConstraint;
+import org.gradle.api.artifacts.ExternalDependency;
+import org.gradle.api.artifacts.ProjectDependency;
+import org.gradle.api.internal.artifacts.dependencies.DefaultProjectDependencyConstraint;
+
+import javax.annotation.Nullable;
+
+/**
+ * Given a declared dependency to be published, determines the coordinates
+ * that should be used to reference the dependency in published metadata. Dependencies
+ * are resolved to component-level precision, meaning the coordinates of the resolved
+ * variant's owning component are returned.
+ */
+public interface ComponentDependencyResolver {
```

Comment:
💭 Is there a more distinct suffix we could give this and the `VariantDepResolver` to keep it conceptually separated from the act of resolving a graph or resolving a dependency?  `Mapper`, `Determiner`, `Lookup`?  Something distinct.

Reply:
Resolver isn't terrible because we use "resolve" to mean "get the thing into this type" a lot of the time.

Reply:
I would prefer to keep the name for now. Messing with the naming of this here is going to cause a ton of merge conflicts with the prototype branch that would be a pain to resolve.